add tests for password and fix an edge case

DeepakRaj228 opened the pull request:
### Summary
Found an invalid edge case where the password method returns a value.

<img width="877" alt="Screenshot 2023-04-05 at 4 57 23 PM" src="https://user-images.githubusercontent.com/30191223/230071392-c37840d3-5cb5-41a9-a045-17d8303e849c.png">

1. Have added an exception if the min length and max length are 0 in the Faker::Internet.password method.
2. Updated the number range starting from 0.
3. Added tests and covered the following existing flows.

> 1) min_length must be at least 2 if mix_case: true && special_characters: false
> Faker::Internet.password(min_length: 10, max_length: 20, mix_case: true) #=> "3k5qS15aNmG"
> 2) min_length must be at least 3 if mix_case: true && special_characters: true
> Faker::Internet.password(min_length: 10, max_length: 20, mix_case: true, special_characters: true) #=> "*%NkOnJsH4"

Test results:
<img width="1607" alt="Screenshot 2023-04-05 at 8 17 42 PM" src="https://user-images.githubusercontent.com/30191223/230117874-b7761cb7-8a52-4a1e-9c8c-c628b340b1ec.png">


<!-- Provide a general description of the code changes in your pull request.

Were there any bugs you had fixed? If so, mention them: Fixes Issue #add-issue-number -->

### Other Information

<!-- If there's anything else that's important and relevant to your pull request, mention that information here.

If you are creating a new generator, share how you are going to use it. Use cases are important to make sure that our faker generators are useful. Also, add `@faker.version next` to help users know when a generator was added. See [this example](https://github.com/faker-ruby/faker/blob/aa31845ed54c25eb2638d716bfddf59771b502aa/lib/faker/music/opera.rb#L68).

Finally, if your pull request affects documentation, please follow the [Guidelines](https://github.com/faker-ruby/faker/blob/main/CONTRIBUTING.md#documentation).

Thanks for contributing to Faker! -->

DeepakRaj228 commented:
@stefannibrasil, I have added a few test cases as you mentioned [here](https://github.com/faker-ruby/faker/issues/2704#issuecomment-1488808056). Please let me know if there are any other cases that need to be covered.

stefannibrasil commented on test/faker/default/test_faker_internet.rb:
```diff
@@ -282,6 +292,24 @@ def test_password_with_invalid_min_length_for_mix_case_and_special_characters
+
+  def test_password_with_invalid_min_length_for_mix_case_only
+    assert_raise_message 'min_length should be at least 2 to enable mix_case configuration' do
+      @tester.password(min_length: 1, mix_case: true, special_characters: false)
```
If we are testing `mix_case` only, then we can remove the `special_characters` from here.

Another suggestion is to assert that it raises an Argument Error with its error message instead of matching the error message only:

```suggestion
      error = assert_raises(ArgumentError) do
      @tester.password(min_length: 1, mix_case: true)
    end
    
    assert_equal 'min_length should be at least 2 to enable mix_case configuration', error.message
```

This way, our test conveys the code behavior more clearly.

stefannibrasil commented on test/faker/default/test_faker_internet.rb:
```diff
@@ -282,6 +292,24 @@ def test_password_with_invalid_min_length_for_mix_case_and_special_characters
   end
 
+  def test_password_with_invalid_min_max_length
+    assert_raise_message 'min_length and max_length must be atleast 1 char in length' do
```
Similar to my comment below (also applies to line 302).

stefannibrasil reviewed (commented):
This is awesome, thank you so much @DeepakRaj228 :pray:

I've left some suggestions/questions, let me know if they make sense.

I think you covered most of them. The only one I can think of now is passing `max_length` only. Then, with it being less  and bigger than min_length.

DeepakRaj228 commented on test/faker/default/test_faker_internet.rb:
```diff
@@ -170,13 +170,16 @@ def test_username_with_range_and_separators
     assert_match(/[a-z]/, password)
     assert_match(/[A-Z]/, password)
+    assert_match(/[0-9]/, password)
+    assert_includes (default_min_length..default_max_length), password.length, 'Password length is incorrect'
```
@stefannibrasil This is the test for a password generated without giving any arguments, the default length variable is set from the password method where we have assigned as min_length: 8 and max_length: 16. Hence I have validated whether the password generated is within the length.

DeepakRaj228 commented:
Hey @stefannibrasil thank you for reviewing this :), I have updated and fixed all your review comments and ran the latest test.

<img width="1610" alt="Screenshot 2023-04-18 at 11 36 48 AM" src="https://user-images.githubusercontent.com/30191223/232686151-dce27490-323a-4e7a-88ac-fe7eec4b83ef.png">

gkunwar commented:
Awesome work @DeepakRaj228. 

I think this PR is ready to merge.

stefannibrasil commented on lib/faker/default/internet.rb:
```diff
@@ -160,6 +160,7 @@ def username(specifier: nil, separators: %w[. _])
       # @faker.version 2.1.3
       def password(min_length: 8, max_length: 16, mix_case: true, special_characters: false)
         raise ArgumentError, 'max_length must be more than min_length' if max_length < min_length
+        raise ArgumentError, 'min_length and max_length must have at least one character' if(min_length == 0 && max_length == 0)
```
ohh, thank you @thdaraujo for catching that! 👀

stefannibrasil commented:
@DeepakRaj228 thank you for your contribution and flexibility. There are only some rubocop issues to solve, and then this can be merged 🎉

stefannibrasil commented:
@DeepakRaj228 you can run rubocop locally to catch the issues before CI: https://github.com/faker-ruby/faker/blob/main/CONTRIBUTING.md#code-style 👀
